src/main/java/com/redhat/rhjmc/containerjfr/net/web/http/api/v2/CertificatePostHandler.java

```diff
@@ -0,0 +1,167 @@
+import com.redhat.rhjmc.containerjfr.net.web.http.api.ApiVersion;
+
+import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;
+import io.vertx.core.http.HttpMethod;
+import io.vertx.ext.web.FileUpload;
+import io.vertx.ext.web.RoutingContext;
+import io.vertx.ext.web.handler.impl.HttpStatusException;
+
+class CertificatePostHandler extends AbstractAuthenticatedRequestHandler {
+
+    static final String PATH = "certificates";
+
+    private final Environment env;
+    private final FileSystem fs;
+    private final Logger logger;
+
+    private static final String TRUSTSTORE_DIR = "TRUSTSTORE_DIR";
+
+    private Function<File, FileOutputStream> outputStreamFunction;
+    private CertificateValidator certValidator;
+
+    @Inject
+    CertificatePostHandler(
+            AuthManager auth,
+            Environment env,
+            FileSystem fs,
+            Logger logger,
+            @Named("OutputStreamFunction") Function<File, FileOutputStream> outputStreamFunction,
+            CertificateValidator certValidator) {
+        super(auth);
+        this.env = env;
+        this.fs = fs;
+        this.logger = logger;
+        this.outputStreamFunction = outputStreamFunction;
+        this.certValidator = certValidator;
+    }
+
+    @Override
+    public ApiVersion apiVersion() {
+        return ApiVersion.V2;
+    }
+
+    @Override
+    public HttpMethod httpMethod() {
+        return HttpMethod.POST;
+    }
+
+    @Override
+    public String path() {
+        return basePath() + PATH;
+    }
+
+    @Override
+    public boolean isAsync() {
+        return false;
+    }
+
+    @Override
+    public boolean isOrdered() {
+        return true;
+    }
+
+    @Override
+    @SuppressFBWarnings("RCN_REDUNDANT_NULLCHECK_WOULD_HAVE_BEEN_A_NPE")
+    public void handleAuthenticated(RoutingContext ctx) throws Exception {
+        FileUpload cert = null;
+        for (FileUpload fu : ctx.fileUploads()) {
+            if ("cert".equals(fu.name())) {
+                cert = fu;
+                break;
+            }
+        }
+
+        if (cert == null) {
+            throw new HttpStatusException(
+                    400, "A file named \"cert\" was not included in the request");
+        }
+
+        Path certPath = fs.pathOf(cert.uploadedFileName());
+
+        if (!env.hasEnv(TRUSTSTORE_DIR)) {
+            throw new HttpStatusException(500, "Truststore directory not set");
+        }
+
+        String truststoreDir = env.getEnv(TRUSTSTORE_DIR);
+        Path filePath = fs.pathOf(truststoreDir, cert.fileName()).normalize();
+
+        if (fs.exists(filePath)) {
+            throw new HttpStatusException(409, filePath.toString() + " Certificate already exists");
+        }
+
+        try (InputStream fis = fs.newInputStream(certPath);
+                FileOutputStream out = outputStreamFunction.apply(filePath.toFile())) {
+
+            Collection certificates = certValidator.parseCertificate(fis);
```

Comment:
Likewise, this `Collection` and its `Iterator` are raw types here but are generics and so should be `Collection<? extends Certificate>` and `Iterator<? extends Certificate>`.

Reply:
So according to the documentation for generateCertificates, the certificate must be DER-encoded, but it can be binary or base64. Therefore, since it seems .pem files are base64 DER-encoded, they also work. I tested it and it seemed to not have any issues saving it, as well as loading it in after a restart. However I can't test that the certificate still works for sure.